File: tutorials/introduction-to-groups-and-communicators/code/groups.c
// Author: Wesley Bland
// Copyright 2015 www.mpitutorial.com
// This code is provided freely with the tutorials on mpitutorial.com. Feel
// free to modify it for your own use. Any distribution of the code must
// either provide a link to www.mpitutorial.com or keep this header intact.
//
// Example using MPI_Comm_split to divide a communicator into subcommunicators
//
#include <stdlib.h>
#include <stdio.h>
#include <mpi.h>

int main(int argc, char **argv) {
  MPI_Init(NULL, NULL);

  // Get the rank and size in the original communicator
  int world_rank, world_size;
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);

  // Get the group of processes in MPI_COMM_WORLD
  MPI_Group world_group;
  MPI_Comm_group(MPI_COMM_WORLD, &world_group);

  int n = 7;
  const int ranks[7] = {1, 2, 3, 5, 7, 11, 13};

  // Construct a group containing all of the prime ranks in world_group
  MPI_Group prime_group;
  MPI_Group_incl(world_group, 7, ranks, &prime_group);

  // Create a new communicator based on the group
  MPI_Comm prime_comm;
  MPI_Comm_create_group(MPI_COMM_WORLD, prime_group, 0, &prime_comm);

  int prime_rank = -1, prime_size = -1;
  // If this rank isn't in the new communicator, it will be MPI_COMM_NULL
  // Using MPI_COMM_NULL for MPI_Comm_rank or MPI_Comm_size is erroneous
  if (MPI_COMM_NULL != prime_comm) {
    MPI_Comm_rank(prime_comm, &prime_rank);
    MPI_Comm_size(prime_comm, &prime_size);
  }

  printf("WORLD RANK/SIZE: %d/%d --- PRIME RANK/SIZE: %d/%d\n",
    world_rank, world_size, prime_rank, prime_size);

  MPI_Group_free(&world_group);
  MPI_Group_free(&prime_group);

  if (MPI_COMM_NULL != prime_comm) {
    MPI_Comm_free(&prime_comm);
  }

  MPI_Finalize();
}

File: tutorials/mpi-fortran/code/groups.f90
program main
  use mpi

  implicit none

  integer            :: world_rank, world_size, ierror
  integer            :: world_group, prime_group
  integer            :: prime_comm
  integer, parameter :: n = 7
  integer            :: ranks(n)
  integer            :: prime_rank, prime_size

  call MPI_Init(ierror)

  ! Get the rank and size in the original communicator
  call MPI_Comm_rank(MPI_COMM_WORLD, world_rank, ierror)
  call MPI_Comm_size(MPI_COMM_WORLD, world_size, ierror)

  ! Get the group of processes in MPI_COMM_WORLD
  call MPI_Comm_group(MPI_COMM_WORLD, world_group, ierror)

  ranks = [1, 2, 3, 5, 7, 11, 13]

  ! Construct a group containing all of the prime ranks in world_group
  call MPI_Group_incl(world_group, 7, ranks, prime_group, ierror)

  ! Create a new communicator based on the group
  call MPI_Comm_create_group(MPI_COMM_WORLD, prime_group, 0, prime_comm, ierror)

  prime_rank = -1
  prime_size = -1
  ! If this rank isn't in the new communicator, it will be MPI_COMM_NULL
  ! Using MPI_COMM_NULL for MPI_Comm_rank or MPI_Comm_size is erroneous
  if (MPI_COMM_NULL .ne. prime_comm) then
    call MPI_Comm_rank(prime_comm, prime_rank, ierror)
    call MPI_Comm_size(prime_comm, prime_size, ierror)
  end if

  print '("WORLD RANK/SIZE: ", I0, "/", I0, " --- PRIME RANK/SIZE: ", I0, "/", I0)', &
        world_rank, world_size, prime_rank, prime_size

  call MPI_Group_free(world_group, ierror)
  call MPI_Group_free(prime_group, ierror)

  if (MPI_COMM_NULL .ne. prime_comm) then
    call MPI_Comm_free(prime_comm, ierror)
  end if

  call MPI_Finalize(ierror)

end program main
